File: Graph/200. Number of Islands/Solution.py
# https://leetcode.com/problems/number-of-islands/

# By Jiapeng
# DFS
# Runtime: 140 ms, faster than 82.64% of Python3 online submissions for Number of Islands.
# Memory Usage: 15.2 MB, less than 5.37% of Python3 online submissions for Number of Islands.

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    count += 1
        return count
        
    def dfs(self, grid, x, y):
        if x >= 0 and y >= 0 and x < len(grid) and y < len(grid[0]) and grid[x][y] == '1':
            grid[x][y] = '0'
            self.dfs(grid, x + 1, y)
            self.dfs(grid, x, y + 1)            
            self.dfs(grid, x - 1, y)            
            self.dfs(grid, x, y - 1)       

File: Graph/200. Number of Islands/Solution.java
class Solution {
    public int numIslands(char[][] grid) {
        if(grid.length==0 || grid[0].length==0)
            return 0;
        int ans = 0;
        for(int i=0; i<grid.length; i++)
            for(int j=0; j<grid[0].length; j++){
                if(grid[i][j]=='1'){
                    dfs(i, j, grid);
                    ans++;
                }
            }
        return ans;
    }
    
    private void dfs(int x, int y, char[][] grid){
        if(x<grid.length && y<grid[0].length && x>=0 && y>=0 && grid[x][y]=='1'){
            grid[x][y] = '0';
            dfs(x+1, y, grid);
            dfs(x, y+1, grid);
            dfs(x-1, y, grid);
            dfs(x, y-1, grid);
            
        }
    }
}
